File: Practice Arrays/12Count max Subarray.py
def countSubarrays(arr, n, k):
    # Write your code here
        n=len(arr)
        left=[0]*n
    
        stack=[]
        for i in range(n):
            while stack and arr[stack[-1]]<k:
                stack.pop()
            left[i]=stack[-1] if stack else -1
            stack.append(i)


        stack.clear()
        right=[0]*n
        # aca considero los iguales pero arriba no ya que ya fueron considerados
        for i in range(n-1,-1,-1):
            while stack and arr[stack[-1]]<=k:
                stack.pop()
            right[i]=stack[-1] if stack else n
            stack.append(i)



        count=0
        for i in range(n):
            if arr[i]==k:
                count+=(i-left[i])*(right[i]-i)
        return count        

if __name__ == "__main__":
    #N, K = list(map(int, input().split()))
    #arr =[2,1,3,4,5,3,2] #list(map(int, input().split()))
    arr=[1,2,1]
    print(countSubarrays(arr, len(arr), 1))

File: Practice Arrays/codigo_cs/12Count max Subarray.cs
using System;
using System.Collections.Generic;

public class Test12
{
    static int CountMaxSubarray(int[] arr, int n, int k)
    {
        // your code goes here
        int[] left = new int[n];
        Stack<int> stack = new();
        for (int i = 0; i < n; i++)
        {
            while (stack.Count > 0 && arr[stack.Peek()] < k)
            {
                stack.Pop();
            }
            left[i] = (stack.Count > 0) ? stack.Peek() : -1;
            stack.Push(i);
        }


        stack.Clear();
        int[] right = new int[n];

        for (int i = n - 1; i >= 0; i--)
        {
            while (stack.Count > 0 && arr[stack.Peek()] <= k)
            {
                stack.Pop();
            }
            right[i] = (stack.Count > 0) ? stack.Peek() : n;
            stack.Push(i);
        }
        int count = 0;
        for (int i = 0; i < n; i++)
        {
            if (arr[i] == k)
            {
                count += (i - left[i]) * (right[i] - i);
            }
        }

        return count;


    }

    public static void Main12()
    {
        int[] data = Console.ReadLine().Split().Select(int.Parse).ToArray();
        int n = data[0];
        int k = data[1];
        int[] arr = Console.ReadLine().Split().Select(int.Parse).ToArray();

        int result = CountMaxSubarray(arr, n, k);
        Console.WriteLine(result);
    }
}
